orchestrator/relayer/src/main_loop.rs

```diff
@@ -23,42 +28,149 @@ pub const ETH_SUBMIT_WAIT_TIME: Duration = Duration::from_secs(600);
 // the donor, while providing maximum utility to the blockchain. Other modes are profitable only
 // and just to always relay everything which is mostly used for tests.
 
-/// If we are relaying in altruistic mode only relay during the lowest 1% of gas prices
-/// in ALTRUISTIC_SAMPLES
-pub const ALTRUISTIC_GAS_PERCENTAGE: f32 = 0.01;
-/// The number of samples over which ALTRUISTIC_GAS_PERCENTAGE is computed, note this is determined
-/// by the relay loop time.
-pub const ALTRUISTIC_SAMPLES: usize = 2000;
+lazy_static! {
+    // Define a gas tracker with a small sample size, the size can be increased later as needed
+    static ref GAS_TRACKER: Arc<RwLock<GasTracker>> = Arc::new(RwLock::new(GasTracker::new(1)));
+    // Tracks the relayer_main_loop's starting point, used to delay batch requests so the gas tracker can store some data
+    static ref RELAYER_START: Instant = Instant::now();
+}
+
+/// fetches the lowest ALTRUISTIC_GAS_PERCENTAGE % of gas prices in the gas tracker's store, use in conjunction with
+/// get_current_gas_price() to determine if now is an acceptable time to relay altruistically
+pub fn get_acceptable_gas_price(percentage: f32) -> Option<Uint256> {
+    let tracker = GAS_TRACKER.try_read();
+    match tracker {
+        Err(_) => None,
+        Ok(gas_tracker) => gas_tracker.get_acceptable_gas_price(percentage),
+    }
+}
+
+/// fetches the last recorded gas price (accurate within relayer_config.gas_tracker_loop_speed seconds) in the
+/// gas tracker's store. Compare against get_acceptable_gas_price() to determine if this is a relatively low gas price
+pub fn get_current_gas_price() -> Option<Uint256> {
+    let tracker = GAS_TRACKER.try_read();
+    match tracker {
+        Err(_) => None,
+        Ok(gas_tracker) => gas_tracker.latest_gas_price(),
+    }
+}
+
+/// writes a new gas price entry into the tracker with the current ethereum gas price
+/// WARNING: only the gas_tracker_loop should call this to track gas uniformly across time.
+/// Adjust relayer_config.gas_tracker_loop_speed and ALTRUISTIC_SAMPLES to control the tracker
+/// panics if the GAS_TRACKER lock is currently held by another thread or is poisoned
+pub async fn update_gas_tracker(web3: &Web3) -> Option<Uint256> {
```

Comment:
consider making this function not public to help enforce the consistency requirements.